airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py

```diff
@@ -227,7 +234,25 @@ def _checkpoint_state(self, stream_name, stream_state, connector_state, logger):
         connector_state[stream_name] = stream_state
         return AirbyteMessage(type=MessageType.STATE, state=AirbyteStateMessage(data=connector_state))
 
+    @lru_cache(maxsize=None)
+    def _get_stream_transformer_and_schema(self, stream_name: str) -> Tuple[Transformer, dict]:
+        """
+        Lookup stream's transform object and jsonschema based on stream name.
+        This function would be called a lot so using caching to save on costly
+        get_json_schema operation.
+        :param stream_name name of stream from catalog.
+        :return tuple with stream transformer object and discover json schema.
+        """
+        stream_instance = self._stream_to_instance_map.get(stream_name)
+        return stream_instance.transformer, stream_instance.get_json_schema()
+
     def _as_airbyte_record(self, stream_name: str, data: Mapping[str, Any]):
         now_millis = int(datetime.now().timestamp()) * 1000
+        transformer, schema = self._get_stream_transformer_and_schema(stream_name)
+        # Transform object fields according to config. Most likely you will
+        # need it to normalize values against json schema. By default no action
+        # taken unless configured. See
+        # docs/connector-development/cdk-python/schemas.md for details.
+        transformer.transform(data, schema)
```

Comment:
I am conflicted between providing this by default in the source itself and therefore increasing the scope of the source vs. just telling the user they are responsible for using this helper class. 

How do you think about this decision?

Reply:
In this case it would be a lot of duplicate code like
```python
class MyStream:
    transformer = TypeTransformer(DefaultTransform)
    def read_records(...):
        for r in super().read_records(...):
            # conversion done in-place, we cant just do yield from ...
             transformer.transform(r)
             yield r
```
instead of just 
```python
class MyStream:
    transformer = TypeTransformer(DefaultTransform)
```
I would like to do this in source code to keep  connectors code  cleaner. Also connector's developer may forget about get_json_schema caching and this  would reduce performance A LOT.